В этом файлике изучение onnx-файлов, в дальнейшем планируется адаптировать библиотеку для работы с ними

### Auxiliary

Всякие вспомогательные функции, взятые из [основного репозитория nanodet](https://github.com/RangiLyu/nanodet?ysclid=mcxcw43rre725659323)

Вроде как, к этой модельке не применимые блинб

In [2]:
from PIL import Image
import numpy as np

_COLORS = (
    np.array(
        [
            0.000,
            0.447,
            0.741,
            0.850,
            0.325,
            0.098,
            0.929,
            0.694,
            0.125,
            0.494,
            0.184,
            0.556,
            0.466,
            0.674,
            0.188,
            0.301,
            0.745,
            0.933,
            0.635,
            0.078,
            0.184,
            0.300,
            0.300,
            0.300,
            0.600,
            0.600,
            0.600,
            1.000,
            0.000,
            0.000,
            1.000,
            0.500,
            0.000,
            0.749,
            0.749,
            0.000,
            0.000,
            1.000,
            0.000,
            0.000,
            0.000,
            1.000,
            0.667,
            0.000,
            1.000,
            0.333,
            0.333,
            0.000,
            0.333,
            0.667,
            0.000,
            0.333,
            1.000,
            0.000,
            0.667,
            0.333,
            0.000,
            0.667,
            0.667,
            0.000,
            0.667,
            1.000,
            0.000,
            1.000,
            0.333,
            0.000,
            1.000,
            0.667,
            0.000,
            1.000,
            1.000,
            0.000,
            0.000,
            0.333,
            0.500,
            0.000,
            0.667,
            0.500,
            0.000,
            1.000,
            0.500,
            0.333,
            0.000,
            0.500,
            0.333,
            0.333,
            0.500,
            0.333,
            0.667,
            0.500,
            0.333,
            1.000,
            0.500,
            0.667,
            0.000,
            0.500,
            0.667,
            0.333,
            0.500,
            0.667,
            0.667,
            0.500,
            0.667,
            1.000,
            0.500,
            1.000,
            0.000,
            0.500,
            1.000,
            0.333,
            0.500,
            1.000,
            0.667,
            0.500,
            1.000,
            1.000,
            0.500,
            0.000,
            0.333,
            1.000,
            0.000,
            0.667,
            1.000,
            0.000,
            1.000,
            1.000,
            0.333,
            0.000,
            1.000,
            0.333,
            0.333,
            1.000,
            0.333,
            0.667,
            1.000,
            0.333,
            1.000,
            1.000,
            0.667,
            0.000,
            1.000,
            0.667,
            0.333,
            1.000,
            0.667,
            0.667,
            1.000,
            0.667,
            1.000,
            1.000,
            1.000,
            0.000,
            1.000,
            1.000,
            0.333,
            1.000,
            1.000,
            0.667,
            1.000,
            0.333,
            0.000,
            0.000,
            0.500,
            0.000,
            0.000,
            0.667,
            0.000,
            0.000,
            0.833,
            0.000,
            0.000,
            1.000,
            0.000,
            0.000,
            0.000,
            0.167,
            0.000,
            0.000,
            0.333,
            0.000,
            0.000,
            0.500,
            0.000,
            0.000,
            0.667,
            0.000,
            0.000,
            0.833,
            0.000,
            0.000,
            1.000,
            0.000,
            0.000,
            0.000,
            0.167,
            0.000,
            0.000,
            0.333,
            0.000,
            0.000,
            0.500,
            0.000,
            0.000,
            0.667,
            0.000,
            0.000,
            0.833,
            0.000,
            0.000,
            1.000,
            0.000,
            0.000,
            0.000,
            0.143,
            0.143,
            0.143,
            0.286,
            0.286,
            0.286,
            0.429,
            0.429,
            0.429,
            0.571,
            0.571,
            0.571,
            0.714,
            0.714,
            0.714,
            0.857,
            0.857,
            0.857,
            0.000,
            0.447,
            0.741,
            0.314,
            0.717,
            0.741,
            0.50,
            0.5,
            0,
        ]
    )
    .astype(np.float32)
    .reshape(-1, 3)
)

def overlay_bbox_cv(img, dets, class_names, score_thresh):
    all_box = []
    for label in dets:
        for bbox in dets[label]:
            score = bbox[-1]
            if score > score_thresh:
                x0, y0, x1, y1 = [int(i) for i in bbox[:4]]
                all_box.append([label, x0, y0, x1, y1, score])
    all_box.sort(key=lambda v: v[5])
    for box in all_box:
        label, x0, y0, x1, y1, score = box
        # color = self.cmap(i)[:3]
        color = (_COLORS[label] * 255).astype(np.uint8).tolist()
        text = "{}:{:.1f}%".format(class_names[label], score * 100)
        txt_color = (0, 0, 0) if np.mean(_COLORS[label]) > 0.5 else (255, 255, 255)
        font = cv2.FONT_HERSHEY_SIMPLEX
        txt_size = cv2.getTextSize(text, font, 0.5, 2)[0]
        cv2.rectangle(img, (x0, y0), (x1, y1), color, 2)

        cv2.rectangle(
            img,
            (x0, y0 - txt_size[1] - 1),
            (x0 + txt_size[0] + txt_size[1], y0 - 1),
            color,
            -1,
        )
        cv2.putText(img, text, (x0, y0 - 1), font, 0.5, txt_color, thickness=1)
    return img

def cv2_imshow(a, convert_bgr_to_rgb=True):
    """A replacement for cv2.imshow() for use in Jupyter notebooks.
    Args:
        a: np.ndarray. shape (N, M) or (N, M, 1) is an NxM grayscale image. shape
            (N, M, 3) is an NxM BGR color image. shape (N, M, 4) is an NxM BGRA color
            image.
        convert_bgr_to_rgb: switch to convert BGR to RGB channel.
    """
    a = a.clip(0, 255).astype('uint8')
    # cv2 stores colors as BGR; convert to RGB
    if convert_bgr_to_rgb and a.ndim == 3:
        if a.shape[2] == 4:
            a = cv2.cvtColor(a, cv2.COLOR_BGRA2RGBA)
        else:
            a = cv2.cvtColor(a, cv2.COLOR_BGR2RGB)
    display(Image.fromarray(a))

COCO_INSTANCE_CATEGORY_NAMES = ['person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
        'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
        'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
        'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
        'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
        'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
        'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard',
        'cell phone', 'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors',
        'teddy bear', 'hair drier', 'toothbrush']

# Тесты

## Информация о модели

In [2]:
import onnx

model_path = 'nanodet_plus_62_192.onnx'
model = onnx.load(model_path)

onnx.checker.check_model(model)  # Проверяем модель на корректность
print(f"Model has {len(model.graph.input)} inputs and {len(model.graph.output)} outputs.")

Model has 1 inputs and 1 outputs.


In [3]:
for input_tensor in model.graph.input:
    print(f"Input name: {input_tensor.name}")
    print(f"Input shape: {input_tensor.type.tensor_type.shape}")
    print(f"Input data type: {input_tensor.type.tensor_type.elem_type}")

Input name: data
Input shape: dim {
  dim_value: 1
}
dim {
  dim_value: 3
}
dim {
  dim_value: 416
}
dim {
  dim_value: 416
}

Input data type: 1


In [4]:
for output_tensor in model.graph.output:
    print(f"Output name: {output_tensor.name}")
    print(f"Output shape: {output_tensor.type.tensor_type.shape}")
    print(f"Output data type: {output_tensor.type.tensor_type.elem_type}")

Output name: output
Output shape: dim {
  dim_value: 1
}
dim {
  dim_value: 3598
}
dim {
  dim_value: 94
}

Output data type: 1


### Первичные выводы

Модель принимает всё вполне логично - изображение в формате (3, 416, 416)

А вот выдаёт что-то непонятное - (1, 3598, 94). Быть может, это изображение M*N, и для каждого пикселя вероятность принадлежности к одному из 94 классов? Посмотрим, какие могут быть M и N.

In [8]:
from functools import reduce

def factors(n):
    return set(reduce(
        list.__add__,
        ([i, n//i] for i in range(1, int(n**0.5) + 1) if n % i == 0)))

display(factors(3598))

{1, 2, 7, 14, 257, 514, 1799, 3598}

Едва ли, поскольку тогда изображение бы было 7x514 или 14x257

Это могло бы быть 3598 найденных объектов, хотя странно, что количество объектов фиксированное, да и что значат 94 признака в таком случае так же неясно. Ведь модель должна быть детектором, то есть возвращать координаты углов и степень уверенности, как это делает изначальный nanodet.

Кстати об изначальном [nanodet](https://github.com/RangiLyu/nanodet?ysclid=mcxcw43rre725659323), вот как он работает:
- возвращает словарь.
- на первом уровне там изображения
- для каждого изображения есть словарь с ключами 0..79, ключ - номер класса.
- внутри содержатся массивы из 5 элементов - координаты четырех углов и уверенность

То есть shape результата оригинального nanodet для одного изображения бы был (1, 79, x, 5), где x - количество найденных объектов для i-го класса

## Исследование конкретных предсказаний модели

Посмотрим, как выглядят предсказания модели для конкретного объекта

In [2]:
import numpy as np
import onnxruntime as ort
import cv2

model_path = 'nanodet_plus_62_192.onnx'
session = ort.InferenceSession(model_path)
input_name = session.get_inputs()[0].name
input_shape = session.get_inputs()[0].shape

img = cv2.imread('testimg.jpg').astype(np.float32) / 255.0
img = cv2.resize(img, (input_shape[2],input_shape[3])).transpose((2, 0, 1))
img = np.expand_dims(img, axis=0)

pred = session.run(None, {input_name: img})
pred[0].shape # (1, 3598, 94)

(1, 3598, 94)

In [3]:
prediction = pred[0]
prediction[0][0]

array([-7.162573  , -7.633009  , -6.9549394 , -6.7796087 , -6.8664594 ,
       -6.9555717 , -7.705238  , -7.0522127 , -6.64586   , -6.885081  ,
       -6.9224095 , -7.4138317 , -6.6935625 , -7.042231  , -6.4374423 ,
       -7.481862  , -7.052189  , -7.132023  , -7.0860085 , -6.313938  ,
       -7.280868  , -6.3987503 , -6.5284295 , -6.595943  , -6.519201  ,
       -6.785137  , -7.203917  , -7.0519457 , -6.8831525 , -7.0770893 ,
       -6.6204333 , -6.863559  , -7.1004205 , -6.3609414 , -6.724513  ,
       -7.042854  , -7.500639  , -7.2029953 , -6.731427  , -6.8650293 ,
       -6.2754393 , -6.21283   , -7.2643843 , -6.584785  , -7.3847322 ,
       -7.1197557 , -7.7141156 , -6.334044  , -6.488992  , -7.050642  ,
       -6.973954  , -6.98145   , -6.568897  , -8.401864  , -6.8272295 ,
       -7.382326  , -6.8288865 , -6.854281  , -6.7770104 , -6.7170777 ,
       -6.244225  , -7.230572  ,  0.37940204, -0.75468206, -0.49067935,
       -0.69336957, -0.7335961 , -0.7304902 , -0.39784607,  1.99

In [26]:
data = prediction[0][0]
print("Среднее значение:", np.mean(data))
print("Дисперсия:", np.var(data))
print("Минимальное значение:", np.min(data))
print("Максимальное значение:", np.max(data))

Среднее значение: -4.6960826
Дисперсия: 10.080989
Минимальное значение: -8.401864
Максимальное значение: 2.4940338


Что-то странное и на вероятности не похоже, если только они зачем-то не были нормализованы в [-10; 10] или какой-то похожий отрезок

Наконец, попробуем зарендерить изображение как будто оно действительно содержит координаты углов и уверенности в первых 5 значениях. Особо на успех не надеемся, но мало ли

In [11]:
from PIL import Image

# Загрузка и предобработка изображения
img = cv2.imread('testimg.jpg').astype(np.float32) / 255.0
input_height = input_shape[2]  # высота входного изображения
input_width = input_shape[3]   # ширина входного изображения
resized_img = cv2.resize(img, (input_width, input_height))  # сохраняем для визуализации
img = resized_img.transpose((2, 0, 1))  # (channels, height, width)
img = np.expand_dims(img, axis=0)       # (1, channels, height, width)

# Получение предсказаний
pred = session.run(None, {input_name: img})[0]  # (1, 3598, 94)

# Обработка предсказаний
threshold = 0  # порог уверенности
pred = pred[0]   # (3598, 94)

for i in range(pred.shape[0]):
    objectness = pred[i, 4]  # уверенность в наличии объекта
    if objectness > threshold:
        # Извлечение координат bounding box (нормализованные)
        bbox = pred[i, 0:4]  # [center_x, center_y, width, height]
        class_probs = pred[i, 5:]  # вероятности классов
        class_id = np.argmax(class_probs)  # индекс класса с максимальной вероятностью
        score = objectness * class_probs[class_id]  # итоговый скор

        # Преобразование в пиксельные координаты
        center_x = bbox[0] * input_width
        center_y = bbox[1] * input_height
        w = bbox[2] * input_width
        h = bbox[3] * input_height
        x_left = int(center_x - w / 2)
        y_top = int(center_y - h / 2)
        x_right = int(center_x + w / 2)
        y_bottom = int(center_y + h / 2)

        # Отрисовка прямоугольника
        cv2.rectangle(resized_img, (x_left, y_top), (x_right, y_bottom), (0, 255, 0), 2)

        # Добавление подписи
        label = f"Class: {class_id}, Score: {score:.2f}"
        cv2.putText(resized_img, label, (x_left, y_top - 10),
                   cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 1)

# Сохранение результата
#Image.fromarray(resized_img).show()
#cv2.imwrite('output_with_predictions.jpg', resized_img)
cv2.imwrite('output_with_predictions.jpg', resized_img*255)

True

Да, там действительно содержится что-то другое